### PR TITLE
improve promxy maintainability

### PR DESCRIPTION
**Pull request by TheoBrigitte**
Aim is to replicate what we re-use from upstream, so we can later on switch to upstream as a droppin replacement and not have to do any refactoring.

Here is are diffs from upstream

```diff
)$ diff <(curl -Ss https://raw.githubusercontent.com/jacksontj/promxy/v0.0.60/pkg/config/config.go) config/config.go
9c9
<       "github.com/jacksontj/promxy/pkg/servergroup"
---
>       "github.com/giantswarm/prometheus-meta-operator/service/controller/resource/promxy/servergroup"
11c11
<       yaml "gopkg.in/yaml.v2"
---
>       "github.com/ghodss/yaml"
```

```diff
$ diff <(curl -Ss https://raw.githubusercontent.com/jacksontj/promxy/v0.0.60/pkg/servergroup/config.go) servergroup/config.go
9c9
<       sd_config "github.com/prometheus/prometheus/discovery/config"
---
>       "github.com/prometheus/prometheus/discovery"
86c86
<       Hosts sd_config.ServiceDiscoveryConfig `yaml:",inline"`
---
>       Hosts discovery.Configs `yaml:",inline"`
```

### CODE REVIEW

File: service/controller/resource/promxy/config/config.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
+package proxyconfig
```
Comment (omegas27) [inline service/controller/resource/promxy/config/config.go:1]: Should that not be package config or promxyconfig?

File: service/controller/resource/promxy/resource.go [hunk 2]
```diff
@@ -99,7 +106,7 @@ func (r *Resource) toServerGroup(cluster metav1.Object) (*ServerGroup, error) {
 				Regex:     relabel.MustNewRegexp(fmt.Sprintf("prometheus-meta-operator;prometheus;%s", key.ClusterID(cluster))),
 			},
 		},
-		KubernetesSDConfigs: []*kubernetes.SDConfig{
+		Hosts: discovery.Configs{
```
Comment (omegas27) [inline service/controller/resource/promxy/resource.go:109]: I could not make it serialize but good if it works :D

File: service/controller/resource/promxy/config/config.go (hunk 1 above)
Comment (TheoBrigitte) [reply service/controller/resource/promxy/config/config.go:1]: that's how its called upstream, weird but I didn't wanted to fix everything.

File: service/controller/resource/promxy/resource.go (hunk 2 above)
Comment (TheoBrigitte) [reply service/controller/resource/promxy/resource.go:109]: That's my only worry here. This have to be tested.

Comment (omegas27) [reply service/controller/resource/promxy/resource.go:109]: I had an issue where Config could not be serialized due to it being an interface. Hopefully, you will figure it out :D

**Comment by TheoBrigitte** [comment]
Can't get any further than this 

```
D 10/07 15:34:08 /apis/cluster.x-k8s.io/v1alpha2/namespaces/default/clusters/4saek promxy checking if promxy configmap needs to be updated | prometheus-meta-operator/service/controller/resource/promxy/delete.go:13 | controller=prometheus-meta-operator-cluster-api-controller | event=delete | loop=1 | version=141817765
D 10/07 15:34:08 /api/v1/namespaces/default/services/kubernetes etcd-certificates created | prometheus-meta-operator/service/controller/resource/generic/create.go:35 | controller=prometheus-meta-operator-control-plane-controller | event=update | loop=0 | version=134430408
D 10/07 15:34:08 /api/v1/namespaces/default/services/kubernetes rbac creating | prometheus-meta-operator/service/controller/resource/rbac/create.go:12 | controller=prometheus-meta-operator-control-plane-controller | event=update | loop=0 | version=134430408
E1007 15:34:08.378094       1 runtime.go:76] Observed a panic: reflect: embedded type with methods not implemented if type is not first field
goroutine 1075 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1787c00, 0x1cac800)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1787c00, 0x1cac800)
        /usr/local/go/src/runtime/panic.go:967 +0x166
gopkg.in/yaml%2ev2.handleErr(0xc000a69618)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:249 +0x9a
panic(0x1787c00, 0x1cac800)
        /usr/local/go/src/runtime/panic.go:967 +0x166
reflect.StructOf(0xc000cd0000, 0xf, 0x11, 0x0, 0x0)
        /usr/local/go/src/reflect/type.go:2477 +0x3149
github.com/prometheus/prometheus/discovery.getConfigType(0x1d365e0, 0x19b5640, 0x0, 0x0)
        /go/pkg/mod/github.com/prometheus/prometheus@v2.21.0+incompatible/discovery/registry.go:103 +0x5ed
github.com/prometheus/prometheus/discovery.UnmarshalYAMLWithInlineConfigs(0x1876760, 0xc000a71560, 0xc000a60fa0, 0x18e3760, 0x16a0501)
        /go/pkg/mod/github.com/prometheus/prometheus@v2.21.0+incompatible/discovery/registry.go:121 +0x149
github.com/giantswarm/prometheus-meta-operator/service/controller/resource/promxy/servergroup.(*Config).UnmarshalYAML(0xc000a71560, 0xc000a60fa0, 0x2b42618, 0x0)
        /root/project/service/controller/resource/promxy/servergroup/config.go:141 +0x7c
gopkg.in/yaml%2ev2.(*decoder).callUnmarshaler(0xc0000a8180, 0xc0004f3500, 0x7fe742466510, 0xc000a71560, 0xc000a71560)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:270 +0xac
gopkg.in/yaml%2ev2.(*decoder).prepare(0xc0000a8180, 0xc0004f3500, 0x1876760, 0xc00020a348, 0x196, 0x1735be0, 0xc0000c8600, 0x1735be0, 0x0)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:313 +0x19c
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc0000a8180, 0xc0004f3500, 0x1876760, 0xc00020a348, 0x196, 0x196)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:364 +0xe3
gopkg.in/yaml%2ev2.(*decoder).sequence(0xc0000a8180, 0xc0004f3490, 0x1737d20, 0xc00017ed00, 0x197, 0x1737d20)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:609 +0x25d
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc0000a8180, 0xc0004f3490, 0x1737d20, 0xc00017ed00, 0x197, 0xc00017ed00)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:374 +0x183
gopkg.in/yaml%2ev2.(*decoder).mappingStruct(0xc0000a8180, 0xc0004f33b0, 0x1857380, 0xc00017ed00, 0x199, 0xc000a69040)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:767 +0xa26
gopkg.in/yaml%2ev2.(*decoder).mapping(0xc0000a8180, 0xc0004f33b0, 0x1857380, 0xc00017ed00, 0x199, 0x1857380)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:626 +0xa37
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc0000a8180, 0xc0004f33b0, 0x1857380, 0xc00017ed00, 0x199, 0xc00017ed00)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:372 +0x1af
gopkg.in/yaml%2ev2.(*decoder).mappingStruct(0xc0000a8180, 0xc000207880, 0x188a7a0, 0xc00017ec30, 0x199, 0x171f5e0)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:767 +0xa26
gopkg.in/yaml%2ev2.(*decoder).mapping(0xc0000a8180, 0xc000207880, 0x188a7a0, 0xc00017ec30, 0x199, 0x188a7a0)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:626 +0xa37
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc0000a8180, 0xc000207880, 0x188a7a0, 0xc00017ec30, 0x199, 0x0)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:372 +0x1af
gopkg.in/yaml%2ev2.(*decoder).document(0xc0000a8180, 0xc000207810, 0x188a7a0, 0xc00017ec30, 0x199, 0x883797)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:384 +0x7c
gopkg.in/yaml%2ev2.(*decoder).unmarshal(0xc0000a8180, 0xc000207810, 0x188a7a0, 0xc00017ec30, 0x199, 0x199)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/decode.go:360 +0x247
gopkg.in/yaml%2ev2.unmarshal(0xc00064e300, 0x6a2, 0x700, 0x1718ea0, 0xc00017ec30, 0x700, 0x0, 0x0)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:148 +0x33f
gopkg.in/yaml%2ev2.Unmarshal(...)
        /go/pkg/mod/gopkg.in/yaml.v2@v2.3.0/yaml.go:81
github.com/giantswarm/prometheus-meta-operator/service/controller/resource/promxy.(*Resource).readFromConfig(0xc00003af80, 0xc000231cc0, 0xc0008032c0, 0x1a3d10d, 0x11)
        /root/project/service/controller/resource/promxy/resource.go:133 +0xf0
github.com/giantswarm/prometheus-meta-operator/service/controller/resource/promxy.(*Resource).EnsureDeleted(0xc00003af80, 0x1d07ce0, 0xc0008032c0, 0x19f7420, 0xc000aec900, 0x18ab680, 0x159828c)
        /root/project/service/controller/resource/promxy/delete.go:20 +0x1c4
github.com/giantswarm/operatorkit/v2/pkg/resource/wrapper/retryresource.(*basicResource).EnsureDeleted.func1(0xc0005f64c0, 0xc000821de0)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/resource/wrapper/retryresource/basic_resource.go:74 +0x64
github.com/cenkalti/backoff.RetryNotify(0xc000a698c0, 0x7fe74260fc80, 0xc0005f64c0, 0xc000a698a0, 0x0, 0x0)
        /go/pkg/mod/github.com/cenkalti/backoff@v2.2.1+incompatible/retry.go:37 +0xb8
github.com/giantswarm/backoff.RetryNotify(0xc000a698c0, 0x1cd3420, 0xc0005f64c0, 0xc000a698a0, 0x192c560, 0x1cd4500)
        /go/pkg/mod/github.com/giantswarm/backoff@v0.2.0/retry.go:21 +0x71
github.com/giantswarm/operatorkit/v2/pkg/resource/wrapper/retryresource.(*basicResource).EnsureDeleted(0xc000606f60, 0x1d07ce0, 0xc0008032c0, 0x19f7420, 0xc000aec900, 0xc0006a8b40, 0xc000252e00)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/resource/wrapper/retryresource/basic_resource.go:86 +0x100
github.com/giantswarm/operatorkit/v2/pkg/resource/wrapper/metricsresource.(*basicResource).EnsureDeleted(0xc00006dc90, 0x1d07ce0, 0xc0008032c0, 0x19f7420, 0xc000aec900, 0x0, 0x0)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/resource/wrapper/metricsresource/basic_resource.go:58 +0x236
github.com/giantswarm/operatorkit/v2/pkg/controller.ProcessDelete(0x1d07ce0, 0xc0008032c0, 0x19f7420, 0xc000aec900, 0xc00028a200, 0x9, 0x10, 0x0, 0x0)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/controller/controller.go:545 +0x1f2
github.com/giantswarm/operatorkit/v2/pkg/controller.(*Controller).deleteFunc(0xc0001dcb60, 0x1d07ce0, 0xc0000aa390, 0x19f7420, 0xc000aec900, 0x9, 0x1d07ce0)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/controller/controller.go:416 +0xc7
github.com/giantswarm/operatorkit/v2/pkg/controller.(*Controller).reconcile(0xc0001dcb60, 0x1d07ce0, 0xc0000aa390, 0xc0000b1fe9, 0x7, 0xc0000b1fe4, 0x5, 0x19f7420, 0xc000aec900, 0x0, ...)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/controller/controller.go:469 +0x368
github.com/giantswarm/operatorkit/v2/pkg/controller.(*Controller).Reconcile(0xc0001dcb60, 0xc0000b1fe9, 0x7, 0xc0000b1fe4, 0x5, 0xc00098e5a0, 0xc000a0c248, 0xc000a0c240, 0xc000a0c240)
        /go/pkg/mod/github.com/giantswarm/operatorkit/v2@v2.0.0/pkg/controller/controller.go:299 +0x4cd
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0001a9c20, 0x1881720, 0xc00000c120, 0x0)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:233 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0001a9c20, 0x203000)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:209 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc0001a9c20)
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:188 +0x2b
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0009fc280)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc0009fc280, 0x1cc92c0, 0xc000654090, 0x1, 0xc000340360)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:156 +0xa3
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc0009fc280, 0x3b9aca00, 0x0, 0x1, 0xc000340360)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.Until(0xc0009fc280, 0x3b9aca00, 0xc000340360)
        /go/pkg/mod/k8s.io/apimachinery@v0.18.8/pkg/util/wait/wait.go:90 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.1/pkg/internal/controller/controller.go:170 +0x411
panic: reflect: embedded type with methods not implemented if type is not first field [recovered]
        panic: reflect: embedded type with methods not implemented if type is not first field [recovered]
        panic: reflect: embedded type with methods not implemented if type is not first field
```

Abandoning this PR.

**Pull request closed without merging**